### PR TITLE
Add `redirect_stdio` (feature from `v1.7`)

### PR DESCRIPTION
**Pull request by MilesCranmer**
From https://github.com/JuliaLang/julia/pull/37978. This completely copies the code from that PR, with one fix for `samefile` -> `Base.Filesystem.samefile`. Unit tests are copied as well.

### CODE REVIEW

**Comment by MilesCranmer** [comment]
Actually wait, I'm being an idiot. I should have just copied it in verbatim – the other functionality added has no effect on existing redirection.

**Comment by MilesCranmer** [comment]
Updated. Now it's a full copy of the code with everything working.

File: src/Compat.jl [hunk 1]
```diff
@@ -650,6 +650,62 @@ if VERSION < v"1.8.0-beta2.17" || v"1.9.0-" <= VERSION < v"1.9.0-DEV.94"
     round(::Type{Bool}, x::AbstractFloat) = (-0.5 <= x < 1.5) ? 0.5 < x : throw(InexactError(:round, Bool, x))
 end
 
+# https://github.com/JuliaLang/julia/pull/37978
+if VERSION < v"1.7.0-beta1"
```
Comment (martinholters) [inline src/Compat.jl:654]: More precisely
```suggestion
if VERSION < v"1.7.0-DEV.1187"
```

**Review by martinholters — COMMENTED** [review]
Needs an entry in the README, but otherwise LGTM. (Also, feel free to bump the version.)

**Comment by MilesCranmer** [comment]
Done!

File: README.md [hunk 2]
```diff
@@ -70,6 +70,8 @@ changes in `julia`.
 
 ## Supported features
 
+* `redirect_stdio`, for simple stream redirection. ([#37978]) (since Compat 4.8.0)
```
Comment (martinholters) [inline README.md:73]: Can you add the link to the list at the end of the file?

Comment (MilesCranmer) [reply README.md:73]: Oops, fixed!

**Pull request merged**